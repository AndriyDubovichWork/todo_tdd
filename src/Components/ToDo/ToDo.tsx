import { Checkbox, FormControlLabel } from '@mui/material';
import React from 'react';

export type TaskType = {
  text: string;
  isCompleted: boolean;
  id: number;
};

type ToDoType = {
  tasks: TaskType[];
  setTasks: React.Dispatch<React.SetStateAction<TaskType[]>>;
};

const ToDo = ({ tasks, setTasks }: ToDoType) => {
  const isCheckedHandler = (id: number, tasks: TaskType[]) => {
    const tasksupdate = tasks.map((task) => {
      if (id === task.id) {
        return {
          id: task.id,
          text: task.text,
          isCompleted: !task.isCompleted,
        };
      }
      return {
        id: task.id,
        text: task.text,
        isCompleted: task.isCompleted,
      };
    });

    setTasks(tasksupdate);
  };
  return (
    <>
      {tasks.map((task: TaskType) => {
        return (
          <FormControlLabel
            key={task.id}
            label={task.text}
            className='column'
            control={
              <Checkbox
                checked={task.isCompleted}
                onChange={() => {
                  isCheckedHandler(task.id, tasks);
                }}
              />
            }
          ></FormControlLabel>
        );
      })}
    </>
  );
};

export default ToDo;
