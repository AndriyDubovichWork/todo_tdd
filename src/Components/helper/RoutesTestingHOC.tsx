import React from 'react';
import { BrowserRouter, Routes, Route } from 'react-router-dom';

type RoutesTestingHOCType = {
  children: JSX.Element;
};
const RoutesTestingHOC = ({ children }: RoutesTestingHOCType) => {
  return (
    <BrowserRouter>
      <Routes>
        <Route path='*' element={children} />
      </Routes>
    </BrowserRouter>
  );
};

export default RoutesTestingHOC;
