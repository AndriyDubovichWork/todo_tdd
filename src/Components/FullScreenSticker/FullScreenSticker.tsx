import React, { useEffect, useState } from 'react';
import { NavLink, useParams, useNavigate } from 'react-router-dom';
import { TaskType } from '../ToDo/ToDo';
import style from './FullScreenSticker.module.scss';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';
import { TaskHandler } from '../helper/TaskHandler';
import { Typography, TextField, Button } from '@mui/material';
const FullScreenSticker = (props: { tasks: TaskType[]; setTasks: any }) => {
  const params = useParams();
  const navigate = useNavigate();

  const [Task, setTask] = useState<TaskType>({
    id: 404,
    text: 'some error',
    isCompleted: false,
  });
  const [isTextCheangeable, setIsTextCheangeable] = useState(false);

  useEffect(() => {
    if (!props.tasks) {
      setTask({
        id: 404,
        isCompleted: false,
        text: 'Some Error',
      });
      return;
    }
    props.tasks.map((task: TaskType) => {
      if (task.id === parseInt(params.toDoId || '404')) {
        setTask(task);
      }
    });
  }, []);

  return (
    <>
      <Typography>
        <NavLink to='/todo_tdd' data-testid={'arrow'}>
          <ArrowBackIcon sx={{ float: 'left', padding: '10px' }} />
        </NavLink>
        <Button
          onClick={() => navigate('/todo_tdd')}
          sx={{ float: 'right', padding: '10px' }}
        >
          save
        </Button>
      </Typography>

      {isTextCheangeable ? (
        <TextField
          label='edit task'
          multiline
          value={Task.text}
          sx={{ width: '100%' }}
          onChange={(e) => {
            TaskHandler(props.tasks, props.setTasks, Task, e.target.value);
          }}
        />
      ) : (
        <Typography
          onDoubleClick={() => {
            setIsTextCheangeable(true);
          }}
          sx={{
            paddingTop: '40px',
            textAling: 'center',
          }}
        >
          {Task.text}
        </Typography>
      )}
    </>
  );
};

export default FullScreenSticker;
