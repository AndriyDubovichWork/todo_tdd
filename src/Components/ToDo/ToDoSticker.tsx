import React, { Dispatch, SetStateAction, useState } from 'react';
import Draggable from 'react-draggable';
import { Box, Checkbox, TextField } from '@mui/material';
import PushPinIcon from '@mui/icons-material/PushPin';
import CloseIcon from '@mui/icons-material/Close';
import OpenInFullIcon from '@mui/icons-material/OpenInFull';
import SaveIcon from '@mui/icons-material/Save';
import { TaskType } from './ToDo';
import LinesEllipsis from 'react-lines-ellipsis';
import { useNavigate } from 'react-router-dom';
import { TaskHandler } from './../helper/TaskHandler';
type ToDoStickerType = {
  task: TaskType;
  tasks: TaskType[];
  isCheckedHandler: (id: number, tasks: TaskType[]) => void;
  setTasks: any;
  Delete: (
    task: TaskType[],
    setTasks: Dispatch<SetStateAction<TaskType[]>>,
    id: number
  ) => void;
};

const ToDoSticker = ({
  task,
  tasks,
  isCheckedHandler,
  Delete,
  setTasks,
}: ToDoStickerType) => {
  const navigate = useNavigate();

  const [isDraggeable, setIsDraggeable] = useState(true);
  const [isTextCheangeable, setIsTextCheangeable] = useState(false);

  const cursor = isDraggeable ? 'grab' : 'auto';
  return (
    <Draggable disabled={!isDraggeable} key={task.id}>
      <Box
        data-testid={'sticker' + task.id}
        style={{
          padding: '10px',
          backgroundColor: '#cece',
          display: 'inline-block',
          overflowWrap: 'break-word',
          wordWrap: 'break-word',
          hyphens: 'auto',
          cursor: cursor,
          width: 200,
          height: 200,
        }}
      >
        <Checkbox
          data-testid='check-box'
          checked={task.isCompleted}
          onChange={() => {
            isCheckedHandler(task.id, tasks);
          }}
          sx={{ cursor: 'pointer' }}
        />

        {!isDraggeable ? (
          <PushPinIcon
            data-testid='pin'
            sx={{ transform: 'rotate(20deg)', cursor: 'pointer' }}
            onClick={() => {
              setIsDraggeable(!isDraggeable);
              setIsTextCheangeable(false);
            }}
          />
        ) : (
          <PushPinIcon
            data-testid='pin'
            sx={{ cursor: 'pointer' }}
            onClick={() => {
              setIsDraggeable(!isDraggeable);
            }}
          />
        )}
        <OpenInFullIcon
          onClick={() => {
            navigate('/todo_tdd/' + task.id);
          }}
          sx={{ cursor: 'pointer' }}
        />
        <CloseIcon
          data-testid={'delete' + task.id}
          sx={{ cursor: 'pointer' }}
          onClick={() => Delete(tasks, setTasks, task.id)}
        />
        {isTextCheangeable ? (
          <TextField
            label='edit task'
            multiline
            rows={5}
            value={task.text}
            onChange={(e) => {
              TaskHandler(tasks, setTasks, task, e.target.value);
            }}
          />
        ) : (
          <LinesEllipsis
            text={task.text}
            onDoubleClick={() => {
              if (isDraggeable) {
                return;
              }
              setIsTextCheangeable(true);
            }}
            maxLine='7'
            ellipsis='...'
            trimRight
            style={{ userSelect: 'none', cursor: cursor }}
            basedOn='letters'
          />
        )}
      </Box>
    </Draggable>
  );
};

export default ToDoSticker;
