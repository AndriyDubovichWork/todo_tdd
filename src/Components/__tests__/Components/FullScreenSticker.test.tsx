import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render, screen, fireEvent } from '@testing-library/react';

import FullScreenSticker from '../../../Components/FullScreenSticker/FullScreenSticker';
import { TaskType } from '../../ToDo/ToDo';
import RoutesTestingHOC from './../../helper/RoutesTestingHOC';
afterEach(() => cleanup());

let tasks: TaskType[] = [
  { text: 'hello 0', id: 0, isCompleted: false },
  { text: 'hello 1', id: 1, isCompleted: true },
  { text: 'hello 2', id: 2, isCompleted: false },
  { text: 'hello 3', id: 3, isCompleted: true },
];

const setTasks = (NewTask: TaskType[]) => {
  tasks = NewTask;
};
describe('FullScreenSticker', () => {
  test('FullScreenSticker have to match snapshot', () => {
    const tree = renderer
      .create(
        <RoutesTestingHOC>
          <FullScreenSticker tasks={tasks} setTasks={setTasks} />
        </RoutesTestingHOC>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('on back arrow click change link', async () => {
    render(
      <RoutesTestingHOC>
        <FullScreenSticker tasks={tasks} setTasks={setTasks} />
      </RoutesTestingHOC>
    );
    const Arrow = screen.getByTestId('arrow');
    await fireEvent.click(Arrow);
    expect(Arrow).toBeInTheDocument();
  });
});
