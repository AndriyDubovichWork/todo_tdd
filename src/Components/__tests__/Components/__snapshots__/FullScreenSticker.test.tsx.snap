// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullScreenSticker FullScreenSticker have to match snapshot 1`] = `
Array [
  <p
    className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
  >
    <a
      className=""
      data-testid="arrow"
      href="/todo_tdd"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-r6b3l7-MuiSvgIcon-root"
        data-testid="ArrowBackIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
        />
      </svg>
    </a>
    <button
      className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-16nntdx-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      save
    </button>
  </p>,
  <p
    className="MuiTypography-root MuiTypography-body1 css-1qyiost-MuiTypography-root"
    onDoubleClick={[Function]}
  >
    some error
  </p>,
]
`;
